internal/network/connection_test.go

```diff
@@ -0,0 +1,241 @@
+)
+
+// mockProxyServer creates a test proxy server that handles CONNECT requests and forwards to target
+func mockProxyServer(t *testing.T, targetHost string, behavior proxyBehavior) (net.Listener, string) {
+	t.Helper()
+
+	proxy := test.NewHTTPServer(t, func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodConnect {
+			t.Errorf("expected CONNECT request, got %s", r.Method)
+			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+			return
+		}
+
+		// Hijack the connection to establish the tunnel
+		hijacker, ok := w.(http.Hijacker)
+		if !ok {
+			t.Fatal("webserver doesn't support hijacking")
+		}
+
+		clientConn, _, err := hijacker.Hijack()
+		if err != nil {
+			t.Fatalf("hijacking failed: %v", err)
+		}
+		defer clientConn.Close()
+
+		switch behavior {
+		case proxyBehaviorBadStatus:
+			// Return a non-200 status
+			clientConn.Write([]byte("HTTP/1.1 403 Forbidden\r\n\r\n"))
+			return
+		case proxyBehaviorNoResponse:
+			// Don't write any response, just close
+			return
+		case proxyBehaviorBadGateway:
+			// Return 502 to simulate proxy returning a bad gateway error
+			clientConn.Write([]byte("HTTP/1.1 502 Bad Gateway\r\n\r\n"))
+			return
+		}
+
+		// Forward the connection to the target
+		targetConn, err := net.Dial("tcp", targetHost)
+		if err != nil {
+			clientConn.Write([]byte("HTTP/1.1 502 Bad Gateway\r\n\r\n"))
+			return
+		}
+		defer targetConn.Close()
+
+		// Send 200 Connection Established
+		clientConn.Write([]byte("HTTP/1.1 200 Connection Established\r\n\r\n"))
+	})
+
+	proxyURL, err := url.Parse(proxy.URL)
+	if err != nil {
+		t.Fatalf("failed to parse proxy URL: %v", err)
+	}
+
+	return proxy.Listener, proxyURL.Host
+}
+
+// mockTargetServer creates a test server that accepts and immediately closes connections
+func mockTargetServer(t *testing.T) (net.Listener, string) {
+	t.Helper()
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("failed to create listener: %v", err)
+	}
+
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				return
+			}
+			// Accept and immediately close the connection
+			conn.Close()
+		}
+	}()
+
+	return listener, listener.Addr().String()
+}
+
+// mockMisbehavingTargetServer creates a test server that does not respond to the connection
+func mockMisbehavingTargetServer(t *testing.T) (net.Listener, string) {
+	t.Helper()
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("failed to create listener: %v", err)
+	}
+	// immediately close the listener to simulate a misbehaving target server
+	listener.Close()
+
+	return listener, listener.Addr().String()
+}
+
+func TestCheckConnectionToHost(t *testing.T) {
+	tests := []struct {
+		name           string
+		useProxy       bool
+		targetServer   serverFunc
+		proxyBehavior  proxyBehavior
+		proxyShouldErr bool
+		wantErr        bool
+		errContains    string // optional substring to check in error message
```

Comment:
these two seem redundant
When we expect an error, we always want to check part of the error message